src/Raven.Server/Documents/Includes/IncludeTimeSeriesCommand.cs

```diff
@@ -0,0 +1,132 @@
+
+        public IncludeTimeSeriesCommand(DocumentDatabase database, DocumentsOperationContext context, IList<string> timeseriesNames, IList<string> fromList, IList<string> toList) 
+            : this(database, context)
+        {
+            _timeSeriesBySourcePath = new Dictionary<string, (IList<string>, IList<string>, IList<string>)>(StringComparer.OrdinalIgnoreCase)
+            {
+                [string.Empty] = (timeseriesNames, fromList, toList)
+            };
+        }
+
+        public IncludeTimeSeriesCommand(DocumentDatabase database, DocumentsOperationContext context, Dictionary<string, HashSet<TimeSeriesRange>> timeSeriesRangesBySourcePath)
+            : this(database, context)
+        {
+            _timeSeriesRangesBySourcePath = timeSeriesRangesBySourcePath;
+        }
+
+        public void Fill(Document document)
+        {
+            if (_timeSeriesRangesBySourcePath != null)
+            {
+                Fill(document, _timeSeriesRangesBySourcePath);
+                return;
+            }
+
+            Fill(document, _timeSeriesBySourcePath);
+        }
+
+
+        private void Fill(Document document, dynamic timeSeriesToGet)
+        {
+            string docId = document.Id;
+
+            foreach (var kvp in timeSeriesToGet)
+            {
+                if (kvp.Key != string.Empty &&
+                    document.Data.TryGet(kvp.Key, out docId) == false)
+                {
+                    throw new InvalidOperationException($"Cannot include time series for related document '{kvp.Key}', " +
+                                                        $"document {document.Id} doesn't have a field named '{kvp.Key}'. ");
+                }
+
+                if (Results.ContainsKey(docId))
+                    continue;
+
+                var rangeResults = GetTimeSeriesForDocument(docId, kvp.Value);
+
+                Results.Add(docId, rangeResults);
+
+            }
+        }
+
+        private List<TimeSeriesRangeResult> GetTimeSeriesForDocument(string docId, HashSet<TimeSeriesRange> timesSeriesToGet)
+        {
+            var rangeResults = new List<TimeSeriesRangeResult>();
+
+            foreach (var range in timesSeriesToGet)
+            {
+                var timeSeriesRangeResult = GetTimeSeries(docId, range.Name, range.From, range.To);
+                rangeResults.Add(timeSeriesRangeResult);
+            }
+
+            return rangeResults;
+        }
+
+        private List<TimeSeriesRangeResult> GetTimeSeriesForDocument(string docId, (IList<string> TimeseriesNames, IList<string> FromList, IList<string> ToList) timesSeriesToGet)
+        {
+            var rangeResults = new List<TimeSeriesRangeResult>();
+
+            for (var i = 0; i < timesSeriesToGet.TimeseriesNames.Count; i++)
+            {
+                var name = timesSeriesToGet.TimeseriesNames[i];
+                var (from, to) = TimeSeriesHandler.ParseDates(timesSeriesToGet.FromList[i], timesSeriesToGet.ToList[i], name);
```

Comment:
It looks like we are parsing those dates all the time? Can this be done once? Maybe lazy?

Reply:
done